share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorShard.java

```diff
@@ -274,79 +272,106 @@ public CoordinatorResult<WriteShareGroupStateResponseData, CoordinatorRecord> wr
+                .setResults(
+                    Collections.singletonList(
+                        WriteShareGroupStateResponse.toResponseWriteStateResult(key.topicId(),
+                            Collections.singletonList(
+                                WriteShareGroupStateResponse.toResponsePartitionResult(
+                                    key.partition()
+                                ))
+                        ))
+                );
+
+            return new CoordinatorResult<>(Collections.singletonList(record), responseData);
+        } catch (IllegalStateException exception) {
+            log.error("Unable to create share state record.", exception);
+            return getWriteErrorResponse(Errors.UNKNOWN_SERVER_ERROR, UNEXPECTED_CURRENT_STATE, key.topicId(), key.partition());
+        }
+    }
 
-        SharePartitionKey key = SharePartitionKey.getInstance(groupId, topicData.topicId(), partitionData.partition());
-        List<CoordinatorRecord> recordList;
-
+    private CoordinatorRecord getRecord(WriteShareGroupStateRequestData.PartitionData partitionData, SharePartitionKey key) {
+        // create appropriate type of record to write to the share group state topic
         if (!shareStateMap.containsKey(key)) {
             // since this is the first time we are getting a write request, we should be creating a share snapshot record
-            recordList = Collections.singletonList(ShareCoordinatorRecordHelpers.newShareSnapshotRecord(
-                groupId, topicData.topicId(), partitionData.partition(), ShareGroupOffset.fromRequest(partitionData)
-            ));
+            // the incoming partition data could have overlapping state batches, we must merge them
+            return ShareCoordinatorRecordHelpers.newShareSnapshotRecord(
+                key.groupId(), key.topicId(), partitionData.partition(),
+                new ShareGroupOffset.Builder()
+                    .setSnapshotEpoch(0)
+                    .setStartOffset(partitionData.startOffset())
+                    .setLeaderEpoch(partitionData.leaderEpoch())
+                    .setStateEpoch(partitionData.stateEpoch())
+                    .setStateBatches(getMergedBatches(Collections.emptyList(), partitionData))
+                    .build());
         } else if (snapshotUpdateCount.getOrDefault(key, 0) >= config.shareCoordinatorSnapshotUpdateRecordsPerSnapshot()) {
-            int newLeaderEpoch = partitionData.leaderEpoch() == -1 ? shareStateMap.get(key).leaderEpoch() : partitionData.leaderEpoch();
-            int newStateEpoch = partitionData.stateEpoch() == -1 ? shareStateMap.get(key).stateEpoch() : partitionData.stateEpoch();
-            long newStartOffset = partitionData.startOffset() == -1 ? shareStateMap.get(key).startOffset() : partitionData.startOffset();
-
+            ShareGroupOffset currentState = shareStateMap.get(key);
+            if (currentState == null) {
```

Comment:
Hmm, this can't happen since we test `!shareStateMap.containsKey(key`) earlier. Ditto in the final `else` clause.

Reply:
added in response to https://github.com/apache/kafka/pull/17149#discussion_r1783239363

Reply:
I had suggested the null check here, I didn't see the `containsKey` above